src/Grpc.AspNetCore.Server.GrpcClient/GrpcHttpClientFactory.cs

```diff
@@ -0,0 +1,115 @@
+using Microsoft.Extensions.Http;
+using Microsoft.Extensions.Options;
+
+namespace Grpc.AspNetCore.Server.GrpcClient
+{
+    internal class GrpcHttpClientFactory<TClient> : INamedTypedHttpClientFactory<TClient> where TClient : ClientBase
+    {
+        private readonly Cache _cache;
+        private readonly IServiceProvider _services;
+        private readonly IHttpContextAccessor _httpContextAccessor;
+        private readonly IOptionsMonitor<GrpcClientOptions> _clientOptions;
+
+        public GrpcHttpClientFactory(Cache cache, IServiceProvider services, IOptionsMonitor<GrpcClientOptions> clientOptions)
+        {
+            if (cache == null)
+            {
+                throw new ArgumentNullException(nameof(cache));
+            }
+
+            if (services == null)
+            {
+                throw new ArgumentNullException(nameof(services));
+            }
+
+            if (clientOptions == null)
+            {
+                throw new ArgumentNullException(nameof(clientOptions));
+            }
+
+            _cache = cache;
+            _services = services;
+            _httpContextAccessor = services.GetService<IHttpContextAccessor>();
+            _clientOptions = clientOptions;
+        }
+
+        public TClient CreateClient(HttpClient httpClient, string name)
+        {
+            if (httpClient == null)
+            {
+                throw new ArgumentNullException(nameof(httpClient));
+            }
+
+            var httpContext = _httpContextAccessor.HttpContext;
+            var serverCallContext = httpContext?.Features.Get<IServerCallContextFeature>().ServerCallContext;
+
+            var namedOptions = _clientOptions.Get(name);
+
+            var callInvoker = new HttpClientCallInvoker(httpClient);
+            if (namedOptions.PropagateCancellationToken)
+            {
+                if (serverCallContext == null)
+                {
+                    throw new InvalidOperationException("Cannot propagate the call cancellation token to the client. Cannot find the current gRPC ServerCallContext.");
+                }
+
+                callInvoker.CancellationToken = serverCallContext.CancellationToken;
+            }
+            if (namedOptions.PropagateDeadline)
+            {
+                if (serverCallContext == null)
+                {
+                    throw new InvalidOperationException("Cannot propagate the call request deadline to the client. Cannot find the current gRPC ServerCallContext.");
+                }
+
+                callInvoker.Deadline = serverCallContext.Deadline;
+            }
+
+            return (TClient)_cache.Activator(_services, new object[] { callInvoker });
+        }
+
+        // The Cache should be registered as a singleton, so it that it can
+        // act as a cache for the Activator. This allows the outer class to be registered
+        // as a transient, so that it doesn't close over the application root service provider.
+        public class Cache
```

Comment:
Why is this public?

Reply:
It's nested in an internal class so it becomes internal anyway.